include/aws/common/atomics_gnu_old.inl

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* These are implicitly included, but help with editor highlighting */
+#include <aws/common/atomics.h>
+#include <aws/common/common.h>
+
+#include <stdint.h>
+#include <stdlib.h>
+
+struct aws_atomic_var {
+    union {
+        size_t intval;
+        void *ptrval;
+    } u;
+};
+
+#define AWS_ATOMIC_INIT_INT_IMPL(x)                                                                                    \
+    {                                                                                                                  \
+        .u = {.intval = (x) }                                                                                          \
+    }
+#define AWS_ATOMIC_INIT_PTR_IMPL(x)                                                                                    \
+    {                                                                                                                  \
+        .u = {.ptrval = (void *)(x) }                                                                                  \
+    }
+
+static inline void aws_atomic_private_compiler_barrier() {
+    __asm__ __volatile__("" : : : "memory");
+}
+
+static inline void aws_atomic_private_barrier_before(enum aws_memory_order order) {
+    if (order == aws_memory_order_release || order == aws_memory_order_acq_rel || order == aws_memory_order_seq_cst) {
+        __sync_synchronize();
```

Comment:
Be careful: this is broken prior to GCC 4.4 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36793)

Reply:
Added a compile-time check that will `#error` on GCC < 4.4